File: ts/src/validate.ts
import { AppTheoryError } from "./errors.js";

export const VALIDATION_RULE_REQUIRED = "required";
export const VALIDATION_RULE_MIN = "min";
export const VALIDATION_RULE_MAX = "max";
export const VALIDATION_RULE_MIN_LENGTH = "min_length";
export const VALIDATION_RULE_MAX_LENGTH = "max_length";
export const VALIDATION_RULE_PATTERN = "pattern";
export const VALIDATION_RULE_ENUM = "enum";

export type ValidationRuleName =
  | typeof VALIDATION_RULE_REQUIRED
  | typeof VALIDATION_RULE_MIN
  | typeof VALIDATION_RULE_MAX
  | typeof VALIDATION_RULE_MIN_LENGTH
  | typeof VALIDATION_RULE_MAX_LENGTH
  | typeof VALIDATION_RULE_PATTERN
  | typeof VALIDATION_RULE_ENUM;

export interface ValidationRuleSpec {
  rule: ValidationRuleName;
  value?: number | string | string[] | undefined;
  field?: string | undefined;
  message?: string | undefined;
}

export interface ValidationFieldError {
  field: string;
  rule: ValidationRuleName;
  message: string;
}

export type ValidationSchema<Req> = Partial<
  Record<Extract<keyof Req, string>, readonly ValidationRuleSpec[]>
>;

export const required = (message?: string): ValidationRuleSpec => ({
  rule: VALIDATION_RULE_REQUIRED,
  message,
});

export const min = (value: number, message?: string): ValidationRuleSpec => ({
  rule: VALIDATION_RULE_MIN,
  value,
  message,
});

export const max = (value: number, message?: string): ValidationRuleSpec => ({
  rule: VALIDATION_RULE_MAX,
  value,
  message,
});

export const minLength = (
  value: number,
  message?: string,
): ValidationRuleSpec => ({
  rule: VALIDATION_RULE_MIN_LENGTH,
  value,
  message,
});

export const maxLength = (
  value: number,
  message?: string,
): ValidationRuleSpec => ({
  rule: VALIDATION_RULE_MAX_LENGTH,
  value,
  message,
});

export const pattern = (
  value: string | RegExp,
  message?: string,
): ValidationRuleSpec => ({
  rule: VALIDATION_RULE_PATTERN,
  value: value instanceof RegExp ? value.source : String(value),
  message,
});

export const oneOf = (
  values: readonly string[],
  message?: string,
): ValidationRuleSpec => ({
  rule: VALIDATION_RULE_ENUM,
  value: [...values],
  message,
});

export function validateValue<Req>(
  value: Req,
  schema: ValidationSchema<Req> | undefined,
): ValidationFieldError[] {
  if (!schema || typeof schema !== "object") return [];
  const record = value as Record<string, unknown>;
  const errors: ValidationFieldError[] = [];
  for (const key of Object.keys(schema)) {
    const rules = schema[key as Extract<keyof Req, string>] ?? [];
    for (const rule of rules) {
      const field = String(rule.field ?? key);
      const error = validateRule(field, record[key], rule);
      if (!error) continue;
      errors.push(error);
      if (rule.rule === VALIDATION_RULE_REQUIRED) break;
    }
  }
  return errors;
}

export function validationError(
  errors: readonly ValidationFieldError[],
): AppTheoryError {
  return new AppTheoryError("app.validation_failed", "validation failed", {
    statusCode: 422,
    details: { errors: errors.map((error) => ({ ...error })) },
  });
}

export function validateOrThrow<Req>(
  value: Req,
  schema: ValidationSchema<Req> | undefined,
): void {
  const errors = validateValue(value, schema);
  if (errors.length > 0) {
    throw validationError(errors);
  }
}

function validateRule(
  field: string,
  value: unknown,
  rule: ValidationRuleSpec,
): ValidationFieldError | null {
  switch (rule.rule) {
    case VALIDATION_RULE_REQUIRED:
      if (isEmptyValue(value)) {
        return fieldError(
          field,
          rule.rule,
          rule.message ?? `${field} is required`,
        );
      }
      return null;
    case VALIDATION_RULE_MIN: {
      const actual = numericValue(value);
      const limit = Number(rule.value);
      if (actual !== null && Number.isFinite(limit) && actual < limit) {
        return fieldError(
          field,
          rule.rule,
          rule.message ?? `${field} must be >= ${String(rule.value)}`,
        );
      }
      return null;
    }
    case VALIDATION_RULE_MAX: {
      const actual = numericValue(value);
      const limit = Number(rule.value);
      if (actual !== null && Number.isFinite(limit) && actual > limit) {
        return fieldError(
          field,
          rule.rule,
          rule.message ?? `${field} must be <= ${String(rule.value)}`,
        );
      }
      return null;
    }
    case VALIDATION_RULE_MIN_LENGTH: {
      const actual = lengthValue(value);
      const limit = Number(rule.value);
      if (actual !== null && Number.isFinite(limit) && actual < limit) {
        return fieldError(
          field,
          rule.rule,
          rule.message ?? `${field} length must be >= ${String(rule.value)}`,
        );
      }
      return null;
    }
    case VALIDATION_RULE_MAX_LENGTH: {
      const actual = lengthValue(value);
      const limit = Number(rule.value);
      if (actual !== null && Number.isFinite(limit) && actual > limit) {
        return fieldError(
          field,
          rule.rule,
          rule.message ?? `${field} length must be <= ${String(rule.value)}`,
        );
      }
      return null;
    }
    case VALIDATION_RULE_PATTERN: {
      const raw = typeof value === "string" ? value : null;
      if (raw === null) return null;
      const patternValue = String(rule.value ?? "");
      if (!new RegExp(patternValue).test(raw)) {
        return fieldError(
          field,
          rule.rule,
          rule.message ?? `${field} must match pattern`,
        );
      }
      return null;
    }
    case VALIDATION_RULE_ENUM: {
      const allowed = Array.isArray(rule.value)
        ? rule.value.map((item) => String(item))
        : String(rule.value ?? "")
            .split("|")
            .map((item) => item.trim())
            .filter(Boolean);
      const actual = String(value ?? "");
      if (!allowed.includes(actual)) {
        return fieldError(
          field,
          rule.rule,
          rule.message ?? `${field} must be one of ${allowed.join(", ")}`,
        );
      }
      return null;
    }
  }
}

function fieldError(
  field: string,
  rule: ValidationRuleName,
  message: string,
): ValidationFieldError {
  return { field, rule, message };
}

function isEmptyValue(value: unknown): boolean {
  if (value === null || value === undefined) return true;
  if (typeof value === "string") return value.trim() === "";
  if (Array.isArray(value)) return value.length === 0;
  if (value instanceof Map || value instanceof Set) return value.size === 0;
  return false;
}

function numericValue(value: unknown): number | null {
  if (typeof value !== "number" || !Number.isFinite(value)) return null;
  return value;
}

function lengthValue(value: unknown): number | null {
  if (typeof value === "string" || Array.isArray(value)) return value.length;
  return null;
}

File: py/src/apptheory/validate.py
from __future__ import annotations

import re
from dataclasses import dataclass
from math import isfinite
from typing import Any

from apptheory.errors import AppTheoryError

VALIDATION_RULE_REQUIRED = "required"
VALIDATION_RULE_MIN = "min"
VALIDATION_RULE_MAX = "max"
VALIDATION_RULE_MIN_LENGTH = "min_length"
VALIDATION_RULE_MAX_LENGTH = "max_length"
VALIDATION_RULE_PATTERN = "pattern"
VALIDATION_RULE_ENUM = "enum"

ValidationRuleName = str


@dataclass(frozen=True, slots=True)
class ValidationRule:
    rule: ValidationRuleName
    value: int | float | str | list[str] | None = None
    field: str = ""
    message: str = ""


@dataclass(frozen=True, slots=True)
class ValidationFieldError:
    field: str
    rule: str
    message: str


ValidationSchema = dict[str, list[ValidationRule]]


def required(message: str = "") -> ValidationRule:
    return ValidationRule(rule=VALIDATION_RULE_REQUIRED, message=message)


def min_value(value: int | float, message: str = "") -> ValidationRule:
    return ValidationRule(rule=VALIDATION_RULE_MIN, value=value, message=message)


def max_value(value: int | float, message: str = "") -> ValidationRule:
    return ValidationRule(rule=VALIDATION_RULE_MAX, value=value, message=message)


def min_length(value: int, message: str = "") -> ValidationRule:
    return ValidationRule(rule=VALIDATION_RULE_MIN_LENGTH, value=value, message=message)


def max_length(value: int, message: str = "") -> ValidationRule:
    return ValidationRule(rule=VALIDATION_RULE_MAX_LENGTH, value=value, message=message)


def pattern(value: str, message: str = "") -> ValidationRule:
    return ValidationRule(rule=VALIDATION_RULE_PATTERN, value=value, message=message)


def one_of(values: list[str] | tuple[str, ...], message: str = "") -> ValidationRule:
    return ValidationRule(rule=VALIDATION_RULE_ENUM, value=[str(value) for value in values], message=message)


def validate_value(value: Any, schema: ValidationSchema | None) -> list[ValidationFieldError]:
    if not schema:
        return []
    errors: list[ValidationFieldError] = []
    for key, rules in schema.items():
        actual = _value_for_key(value, key)
        for rule in rules:
            field = str(rule.field or key)
            field_error = _validate_rule(field, actual, rule)
            if field_error is None:
                continue
            errors.append(field_error)
            if rule.rule == VALIDATION_RULE_REQUIRED:
                break
    return errors


def validate_or_raise(value: Any, schema: ValidationSchema | None) -> None:
    errors = validate_value(value, schema)
    if errors:
        raise validation_error(errors)


def validation_error(errors: list[ValidationFieldError]) -> AppTheoryError:
    return AppTheoryError(
        code="app.validation_failed",
        message="validation failed",
        status_code=422,
        details={"errors": [{"field": err.field, "rule": err.rule, "message": err.message} for err in errors]},
    )


def _validate_rule(field: str, value: Any, rule: ValidationRule) -> ValidationFieldError | None:
    config_error = _validate_rule_config(field, rule)
    if config_error is not None:
        return config_error

    if rule.rule == VALIDATION_RULE_REQUIRED:
        if _is_empty(value):
            return _field_error(field, rule, rule.message or f"{field} is required")
        return None
    if rule.rule in {VALIDATION_RULE_MIN, VALIDATION_RULE_MAX}:
        return _validate_numeric_rule(field, value, rule)
    if rule.rule in {VALIDATION_RULE_MIN_LENGTH, VALIDATION_RULE_MAX_LENGTH}:
        return _validate_length_rule(field, value, rule)
    if rule.rule == VALIDATION_RULE_PATTERN:
        return _validate_pattern_rule(field, value, rule)
    if rule.rule == VALIDATION_RULE_ENUM:
        return _validate_enum_rule(field, value, rule)
    return None


def _validate_numeric_rule(field: str, value: Any, rule: ValidationRule) -> ValidationFieldError | None:
    actual = _numeric(value)
    limit = _rule_number(rule.value)
    if actual is None or limit is None:
        return None
    if rule.rule == VALIDATION_RULE_MIN and actual < limit:
        return _field_error(field, rule, rule.message or f"{field} must be >= {rule.value}")
    if rule.rule == VALIDATION_RULE_MAX and actual > limit:
        return _field_error(field, rule, rule.message or f"{field} must be <= {rule.value}")
    return None


def _validate_length_rule(field: str, value: Any, rule: ValidationRule) -> ValidationFieldError | None:
    actual = _length(value)
    limit = _rule_number(rule.value)
    if actual is None or limit is None:
        return None
    if rule.rule == VALIDATION_RULE_MIN_LENGTH and actual < limit:
        return _field_error(field, rule, rule.message or f"{field} length must be >= {rule.value}")
    if rule.rule == VALIDATION_RULE_MAX_LENGTH and actual > limit:
        return _field_error(field, rule, rule.message or f"{field} length must be <= {rule.value}")
    return None


def _validate_pattern_rule(field: str, value: Any, rule: ValidationRule) -> ValidationFieldError | None:
    if isinstance(value, str) and re.search(str(rule.value or ""), value) is None:
        return _field_error(field, rule, rule.message or f"{field} must match pattern")
    return None


def _validate_enum_rule(field: str, value: Any, rule: ValidationRule) -> ValidationFieldError | None:
    allowed = _enum_values(rule.value)
    if str(value or "") not in allowed:
        return _field_error(field, rule, rule.message or f"{field} must be one of {', '.join(allowed)}")
    return None


def _validate_rule_config(field: str, rule: ValidationRule) -> ValidationFieldError | None:
    invalid = False
    if rule.rule == VALIDATION_RULE_REQUIRED:
        invalid = rule.value is not None and str(rule.value).strip() != ""
    elif rule.rule in {VALIDATION_RULE_MIN, VALIDATION_RULE_MAX}:
        invalid = _rule_number(rule.value) is None
    elif rule.rule in {VALIDATION_RULE_MIN_LENGTH, VALIDATION_RULE_MAX_LENGTH}:
        number = _rule_number(rule.value)
        invalid = number is None or int(number) != number
    elif rule.rule == VALIDATION_RULE_PATTERN:
        try:
            re.compile(str(rule.value or ""))
        except re.error:
            invalid = True
    elif rule.rule == VALIDATION_RULE_ENUM:
        invalid = not _enum_values(rule.value)
    else:
        invalid = True
    if not invalid:
        return None
    return _field_error(
        field,
        rule,
        rule.message or f"{field} has invalid validation rule {rule.rule}",
    )


def _field_error(field: str, rule: ValidationRule, message: str) -> ValidationFieldError:
    return ValidationFieldError(field=field, rule=str(rule.rule), message=message)


def _value_for_key(value: Any, key: str) -> Any:
    if isinstance(value, dict):
        return value.get(key)
    return getattr(value, key, None)


def _is_empty(value: Any) -> bool:
    return value is None


def _numeric(value: Any) -> float | None:
    if isinstance(value, bool):
        return None
    if isinstance(value, int | float):
        return float(value)
    return None


def _rule_number(value: Any) -> float | None:
    if value is None or isinstance(value, bool) or str(value).strip() == "":
        return None
    try:
        out = float(str(value))
    except ValueError:
        return None
    return out if isfinite(out) else None


def _enum_values(value: Any) -> list[str]:
    allowed = [str(item) for item in value] if isinstance(value, list) else str(value or "").split("|")
    return [item.strip() for item in allowed if item.strip()]


def _length(value: Any) -> int | None:
    if isinstance(value, str | list | tuple | set | dict):
        return len(value)
    return None
